### DISCUSSION TITLE
Implementing Single Object of DB Instance in NextJS

### DISCUSSION THREAD

**Discussion by serkan-bayram** [pocketbase/pocketbase#4135] [Q&A] [answered]
I'm trying to implement this code but it turns out Next has a couple of problems with top level await (https://github.com/vercel/next.js/issues/54282). While waiting, is there any workarounds that can solve this?

The code I'm trying to implement:
```js
import PocketBase from "pocketbase";

const pb = new PocketBase(process.env.dbURL);

await pb.admins.authWithPassword(
  process.env.dbUsername,
  process.env.dbPassword,
  {
    // This will trigger auto refresh or auto reauthentication in case
    // the token has expired or is going to expire in the next 30 minutes.
    autoRefreshThreshold: 30 * 60,
  }
);

export default pb;
```

**Comment by jonshipman** [comment]
I'd wrap that in a function at the very least. I assume your intention is to give admin level auth to your entire application, however, that can be tricky if it's not explicit.

Personally I favor having two functions, e.g. `getPocketbase` and `getPocketbaseAsAdmin`. 

To answer your question, NextJS does have some caching written into the global fetch. You will need to disable that to get auth to work in Next.

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand the root of the problem, but if you can't use top level `await` then, similarly as the other user suggested, you can try to wrap it in async function and await it inside from your components. Something like:

```js
// /src/adminClient.js
import PocketBase from "pocketbase";

let pb;

async function getAdminClient() {
  if (!pb) {
    pb = new PocketBase(process.env.dbURL);

    // maybe also disable auto cancellation for the admin client
    // in case of concurrent requests from different component handlers
    pb.autoCancellation(false)

    await pb.admins.authWithPassword(
      process.env.dbUsername,
      process.env.dbPassword,
      {
        // This will trigger auto refresh or auto reauthentication in case
        // the token has expired or is going to expire in the next 30 minutes.
        autoRefreshThreshold: 30 * 60,
      }
    );
  }

  return pb;
}

export default getAdminClient;
```

Then from your components you can import it like this:

```js
import getAdminClient from "./src/adminClient.js"

...

const adminClient = await getAdminClient()

const result = await adminClient.collection("example").getList(1, 50)
```

**Comment by serkan-bayram** [comment]
I've been dealing with this problem for a while now, I will write everything in detail but I don't think it's even related to pocketbase anymore so feel free to not answer if you don't think of any new solutions. 

First of all this was my first code and it was working fine, but there is one problem: It may work slow due to encryption stuff.

```js
// The function that I call every time I want to make a DB Connection.
export async function authAsAdmin() {
  const pb = new PocketBase(process.env.dbURL);

  await pb.admins.authWithPassword(
    process.env.dbUsername,
    process.env.dbPassword,
  );

  return pb;
}
```

So, to solve this slowness issue you sugessted to use single global server instance, and I tried that.

```js
// src/adminClient.js
import PocketBase from "pocketbase"

const adminClient = new PocketBase(process.env.dbURL);

await adminClient.admins.authWithPassword(process.env.dbUsername, process.env.dbPassword, {
  // This will trigger auto refresh or auto reauthentication in case
  // the token has expired or is going to expire in the next 30 minutes.
  autoRefreshThreshold: 30 * 60
})

export default adminClient;
```

When I try to use the code it actually works great, except in "server actions". Due to some bug or something else top level awaits are not working correctly in Nextjs server actions. (https://github.com/vercel/next.js/issues/54282)

Because of the bug I couldn't use top level await and as your suggestion I tried to wrap it in an async function:

```js

import PocketBase from "pocketbase";

let pb;

export async function authAsAdmin() {
  if (!pb) {
    pb = new PocketBase(process.env.dbURL);

    pb.autoCancellation(false);

    await pb.admins.authWithPassword(
      process.env.dbUsername,
      process.env.dbPassword,
      {
        // This will trigger auto refresh or auto reauthentication in case
        // the token has expired or is going to expire in the next 30 minutes.
        autoRefreshThreshold: 30 * 60,
      }
    );
  }

  return pb;
}
```

Now, in theory this code should work as same as top level await but it doesn't. Somehow the first code that tries to reach this function does not work correctly. This is the first code on one of my pages and here is it's output:

```js

import { authAsAdmin } from "./authAsAdmin";

export async function getMenu() {
  try {
    const pb = await authAsAdmin();

    console.log(pb);

    const record = await pb
      .collection("menus")
      .getFirstListItem(null, { sort: "-created" });

    return record;
  } catch (error) {
    console.log("Error on getMenu: ", error);
    return null;
  }
}


```

Output: 

```

<ref *1> Client {
  cancelControllers: {},
  recordServices: {},
  enableAutoCancellation: false,
  baseUrl: 'http://myip/',
  lang: 'en-US',
  authStore: LocalAuthStore {
    baseToken: '',
    baseModel: null,
    _onChangeCallbacks: [],
    storageFallback: {},
    storageKey: 'pocketbase_auth'
  },
  admins: AdminService { client: [Circular *1] },
  collections: CollectionService { client: [Circular *1] },
  files: FileService { client: [Circular *1] },
  logs: LogService { client: [Circular *1] },
  settings: SettingsService { client: [Circular *1] },
  realtime: RealtimeService {
    client: [Circular *1],
    clientId: '',
    eventSource: null,
    subscriptions: {},
    lastSentTopics: [],
    maxConnectTimeout: 15000,
    reconnectAttempts: 0,
    maxReconnectAttempts: Infinity,
    predefinedReconnectIntervals: [
       200,  300,  500,
      1000, 1200, 1500,
      2000
    ],
    pendingConnects: []
  },
  health: HealthService { client: [Circular *1] },
  backups: BackupService { client: [Circular *1] }
}
Error on getMenu:  ClientResponseError 403: Only admins can perform this action.
    at eval (webpack-internal:///(rsc)/./node_modules/pocketbase/dist/pocketbase.es.mjs:1069:23)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async getMenu (webpack-internal:///(rsc)/./app/components/Functions/getMenu.jsx:12:24)
    at async getOtherRatings (webpack-internal:///(rsc)/./app/oyla/components/getOtherRatings.jsx:13:32)
    at async Status (webpack-internal:///(rsc)/./app/components/Status.jsx:16:22) {
  url: 'http://myip/api/collections/menus/records?page=1&perPage=1&skipTotal=1&sort=-created',
  status: 403,
  response: {
    code: 403,
    message: 'Only admins can perform this action.',
    data: {}
  },
  isAbort: false,
  originalError: {
    url: 'http://myip/api/collections/menus/records?page=1&perPage=1&skipTotal=1&sort=-created',
    status: 403,
    data: {
      code: 403,
      message: 'Only admins can perform this action.',
      data: {}
    }
  }
}

```

As you can see it indicates that I'm somehow not admin but I don't think this is the case, for some reason pb instance is kinda empty and after these two inputs "pb instance" prints out a couple of times and they look just fine.

So, I thought to intiliaze database connection in somewhere else before these codes try to reach and I tried to call "authAsAdmin" in my middleware and RootLayout, none of them worked. 

Long story short I think I will go with the slow function until top level await got fixes and then I can try to update my codebase.

**Reply by ganigeorgiev** [reply]
> Long story short I think I will go with the slow function until top level await got fixes and then I can try to update my codebase.

I don't recommend doing this as calling `authWithPassword()` on every request will be terribly slow. 
I'd advice to make sure that there is no caching involved, please read [Next.js-Opting out of data caching](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#opting-out-of-data-caching) (so far almost all recent Next.js questions were cache related).

**Reply by serkan-bayram** [reply]
Yes, it finally worked, and it might be cache related. I have no idea what is going on here but this the code for one of my pages:

```js
export default async function Page() {
  const headersList = headers();
  const isGuest = headersList.get("is-guest");

  const menu = await getMenu();
  
  ...

  return (
      ...
  );
}
```

And deleting these headers lines fixed my problem. 

```js
export default async function Page() {
  const menu = await getMenu();
  
  ...

  return (
      ...
  );
}
```

Here is the last version of my single global instance incase of any future readers: 

```js
import PocketBase from "pocketbase";

let pb;

export async function authAsAdmin() {
  if (!pb) {
    pb = new PocketBase(process.env.dbURL);

    pb.autoCancellation(false);

    await pb.admins.authWithPassword(
      process.env.dbUsername,
      process.env.dbPassword,
      {
        // This will trigger auto refresh or auto reauthentication in case
        // the token has expired or is going to expire in the next 30 minutes.
        autoRefreshThreshold: 30 * 60,
      }
    );
  }

  return pb;
}

```

Thanks so much for answering patiently. Have a good day.
